apps/emqx_retainer/src/emqx_retainer_index.erl

```diff
@@ -0,0 +1,177 @@
+%% according index key.
+%%
+%% E.g. for `[2,3]' index and <code>['+', <<"b">>, '+', <<"d">>]</code> wildcard topic
+%% the score is `1', because the according index key pattern is
+%% <code>{[<<"b">>, '_'], ['_', <<"d">>]}</code>.
+%%
+%% @see foreach_index_key/3
+%% @see to_index_key/2
+-spec index_score(index(), emqx_topic:words()) -> non_neg_integer().
+index_score(Index, Tokens) ->
+    index_score(Index, Tokens, 1, 0).
+
+
+%% @doc Given a list of indices and a wildcard topic
+%% returns index with the best score.
+%%
+%% Returns `undefined' if there are no indices with score `> 0'.
+%%
+%% @see index_score/2
+-spec select_index(emqx:words(), list(index())) -> index() | undefined.
+select_index(Tokens, Indices) ->
+    select_index(Tokens, Indices, 0, undefined).
+
+%% @doc For an index and a wildcard topic
+%% returns a matchspec pattern for the corresponding index key.
+%%
+%% E.g. for `[2, 3]' index and <code>['+', <<"b">>, '+', <<"d">>]</code> wildcard topic
+%% returns <code>{[2, 3], {[<<"b">>, '_'], ['_', <<"d">>]}}</code> pattern.
+-spec condition(index(), emqx_topic:words()) -> match_pattern_part().
+condition(Index, Tokens) ->
+    {Index, condition(Index, Tokens, 1, [], [])}.
+
+%% @doc Returns a matchspec pattern for a wildcard topic.
+%%
+%% E.g. for <code>['+', <<"b">>, '+', <<"d">>, '#']</code> wildcard topic
+%% returns <code>['_', <<"b">>, '_', <<"d">> | '_']</code> pattern.
+-spec condition(emqx_topic:words()) -> match_pattern_part().
+condition(Tokens) ->
+    Tokens1 = [case W =:= '+' of true -> '_'; _ -> W end || W <- Tokens],
+    case length(Tokens1) > 0 andalso lists:last(Tokens1) =:= '#' of
+        false -> Tokens1;
+        _ -> (Tokens1 -- ['#']) ++ '_'
+    end.
+
+%% @doc Restores concrete topic from its index key representation.
+%%
+%% E.g given `{[2, 3], {[<<"b">>, <<"c">>], [<<"a">>, <<"d">>]}}' index key
+%% returns `[<<"a">>, <<"b">>, <<"c">>, <<"d">>]' topic.
+-spec restore_topic(index_key()) -> emqx_topic:words().
+restore_topic({Index, {IndexTokens, OtherTokens}}) ->
+    restore_topic(Index, IndexTokens, OtherTokens, 1, []).
+
+%%--------------------------------------------------------------------
+%% Private
+%%--------------------------------------------------------------------
+
+split_index_tokens([NIndex | OtherIndex], [Token | Tokens], N, IndexTokens, OtherTokens)
+    when NIndex == N ->
+    split_index_tokens(OtherIndex, Tokens, N + 1, [Token | IndexTokens], OtherTokens);
+split_index_tokens([_NIndex | _] = Index, [Token | Tokens], N, IndexTokens, OtherTokens) ->
+    split_index_tokens(Index, Tokens, N + 1, IndexTokens, [Token | OtherTokens]);
+split_index_tokens([], Tokens, _N, IndexTokens, OtherTokens) ->
+    {lists:reverse(IndexTokens), lists:reverse(OtherTokens) ++ Tokens};
+split_index_tokens(_Index, [], _N, IndexTokens, OtherTokens) ->
+    {lists:reverse(IndexTokens), lists:reverse(OtherTokens)}.
+
+index_score([N | _Index], [Ph | _Tokens], N, Score)
```

Comment:
need some docs to explain what a score is

Reply:
There are some docs for `index_score/2` public function.